### DISCUSSION TITLE
Client side rendering (wp-each and other directives)

### DISCUSSION THREAD

**Discussion by gaambo** [WordPress/gutenberg#58572] [Interactivity API]
Let me prefix everything here with: I initially was skeptical whether the interactivit API would be a good addition to WordPress, since there are already things like AlpineJS wildly used in the web dev community. That's why I'm trying to build things with the interactivity API to get a better feeling for it. And I have to say so far, that I can really imagine using it for things like filterable lists, querying, client side navigation etc. I know that this API is still in an early phase and experimental, but what I see so far seems well-built :) Props to all contributors who worked on this! 👍

I'm playing around with the new `wp-each` directive (just released in Gutenberg 17.6) to render a list of filtered objects (CPTs) completely on the client side and also handle filtering and re-rendering on the client side. 
I've previously built things like that with AlpineJS, so it may very well be, that I'm still using the interactivity API wrong or am using the wrong directives. 
I've got some questions, ideas and problems and thought I'd collect them here in this thread, but if this is not the right place or separate issues are better, just let me know. (Maybe there are already existing issues and I just couldn't find them). 

## Evaluating the contents of attributes directly
AlpineJS has the handy possibility to directly evaluate the contents of attributes, without having to use a callback function on the store. 
For example: 
`<span data-wp-class--is-active="context.currentFilter = context.filter.id" data-wp-text="context.filter.name"></span>`
instead of having to use
`<span data-wp-class--is-active="callbacks.isCurrentFilter" data-filter="context.filter.id" data-wp-text="context.filter.name"></span>`

(This runs in a data-wp-each where context.filter is an item in a context.filters list)

I don't know how Alpine JS does it and if they are using `eval` for that and how that could be a security problem. I guess developers need to be aware, that any user values in the attribute definitely should be escaped.

This is also handy to call helper functions. For example, formatting a date in wp-text. I guess this could be done via a computed/derived value (via a getter). But since the value is from an object in an array (which is part of the wp-each list), I think a getter on the store would be kind of "misplaced". 

Most of the examples just use boolean values (toggle, isToggled, etc.), but do not compare any values.

## Passing arguments to callbacks
Following the example above, to conditionally add the `is-active` class on elements, I need to use a callback. But as far as I can see, there's no way to pass arguments to a callback being called for a directive. The only argument passed to the callback is an event. I can use the event's target **or** `getElement` and then get the elements data-attribute to check. But this is kind of cumbersome: 
```js
isCurrentFilter: () => {
      const { ref: button } = getElement();
      const { currentFilter } = getContext();
      if (!currentFilter || !button) {
        return false;
      }
      const filter = button.dataset.filter;
      return filter && filter === currentFilter;
    },
```

Or is there an existing better way? 

## getElement ref is null on init
In the above example, when setting the `currentFilter` value in `data-wp-context` on the server side, `ref` is null on the first runs of `isCurrentFilter` for all objects. So I can't use the data-attribute value. If I return false here, the isCurrentFilter is not evaluated again, only when a context value changes. I guess this is, because the first run runs before the element is rendered?

## Hide/show elements (x-if/x-show attributes).
The examples use the `hidden` attribute to conditionally hide elements (e.g.: `data-wp-bind--hidden="callback.isHidden"`)
Alpine JS has a handy `x-show` attribute, to only render an element if the attribute evaluates to true. It just (visually) hides the element, like the hidden property. `x-if` can be used on template-tags to conditionally _render_ the element. 
It would be great to have something like that in the Interactivity API. 

An example: 
```html<
  <span data-wp-show="!!context.event.liveLink"><svg [...] ></svg> <a data-wp-bind--href="context.event.liveLink">Watch Live</a></span>
  <span data-wp-show="!context.event.liveLink">This event is not streamed live</span>
```
At the moment, I have to do: 
```html
<span data-wp-bind--hidden="callbacks.hasNoLiveLink"><svg [...] ></svg> <a data-wp-bind--href="context.event.liveLink">Watch Live</a></span>
<span data-wp-bind--hidden="callbacks.hasLiveLink">This event is not streamed live</span>
```
Which is kind of confusing to read (because of the inverse negations) and I also have to use callbacks (see first idea above). I like to use the hidden attribute (or x-show), because it also semantically transports that the element is hidden / not relevant etc. - just applying a CSS class (e.g. `data-wp-class--is-hidden` could be a workaround, but not perfect IMHO).

I know this may all be hard to understand without context, so I uploaded a simplified version of my template in a gist (this is kind of pseudocode because I simplified it, so not sure if you can copy-paste that and it would work 😅): 
https://gist.github.com/gaambo/d9803da55d03a5f9bea3875f9eaa3acc

As mentioned above, I'd like this thread to serve as a discussion for ideas how those things could be done or which new directives could help make it easier. All of this are things that are especially useful when client side rendering (e.g. inside a wp-each), but also when setting values from the server side.
I'm happy for any hints, feedback and tips :)

**Comment by luisherranz** [comment]
> if this is not the right place or separate issues are better, just let me know

This is the perfect place 🙂

> Evaluating the contents of attributes directly
> [...]
> I don't know how Alpine JS does it and if they are using eval for that and how that could be a security problem.

Exactly, we didn't want to support that to avoid opening a can of security problems.

Also, it's not only about your own code. If a block uses that syntax, all the sites that use that block can't enable a strong CSP anymore, which could lead to exploits in other areas of your site.

We do support a single negation operator, though, because that's safe to read/parse and very handy:

```html
<span data-wp-bind--hidden="!context.isOpen">Opened!</span>
<span data-wp-bind--hidden="context.isOpen">Closed!</span>
```

> This is also handy to call helper functions. For example, formatting a date in wp-text. I guess this could be done via a computed/derived value (via a getter). But since the value is from an object in an array (which is part of the wp-each list), I think a getter on the store would be kind of "misplaced".

That's a perfectly valid usage of "derived state". Basically, it would work like this:

```html
<template data-wp-each--filter="state.filters">
  <span
    data-wp-class--is-active="state.isCurrentFilter"
    data-filter="context.filter.id"
    data-wp-text="context.filter.name"
  ></span>
</template>
```

```js
store("myPlugin", {
  state: {
    filters: [],
    get isCurrentFilter() {
      const { filter } = getContext();
      return filter.id === filter.name;
    },
  },
});
```

> getElement ref is null on init

Yes. `ref` is basically the result of `useRef` in (P)React, so it's null the first time the element is about to be rendered.

If you want to access the element attributes, you can use the `attributes` object (previously known as `props`) which are always defined.

```js
store("myPlugin", {
  state: {
    get isCurrentFilter() {
      const { attributes } = getElement();
      const { currentFilter } = getContext();
      return attributes["data-filter"] === currentFilter;
    },
  },
});
```

> But as far as I can see, there's no way to pass arguments to a callback being called for a directive.

You can use attributes (as in the example above) or context, whatever suits your case better.

Same example with context instead of attributes:

```html
<div data-wp-context='{ "filter": "some filter" }'></div>
```

```js
store("myPlugin", {
  state: {
    get isCurrentFilter() {
      const { currentFilter, filter } = getContext();
      return filter === currentFilter;
    },
  },
});
```

> Hide/show elements (x-if/x-show attributes).

We had that at the very beginning but it was not clear how to do the server directive processing so removed it in favor of `data-wp-bind--hidden` and `data-wp-style--display`.

We still want to add it though, because it has some benefits in some cases.

There is a conversation about how it should work here, in case you want to participate:

- https://github.com/WordPress/gutenberg/discussions/53058

> I also have to use callbacks (see first idea above)

You can use the negation operator to avoid one piece of state:

```html
<span data-wp-bind--hidden="!state.hasLiveLink">...</span>
<span data-wp-bind--hidden="state.hasLiveLink">...</span>
```

By the way, even though callbacks work fine in the client (any function works), you should also use derived state (getters) instead of callback functions when passing state to the directives.

```js
store("myPlugin", {
  state: {
    get hasLiveLink() {
      const { context } = getContext();
      return !!context.event.liveLink;
    },
  },
});
```

Then, when you need to use it in another action/derived-state, it's as if it was a piece of state.

```js
const { state } = store("myPlugin", {
  state: {
    get hasLiveLink() {
      const { context } = getContext();
      return !!context.event.liveLink;
    },
  },
  actions: {
    setLiveLink() {
      if (state.hasLiveLink) {
        // do something...
      } else {
        // do something else...
      }
    },
  },
});
```

It's also a bit more performant because underneath we are wrapping those getters with Preact's `computed`, and it can be populated in the server (unlike callbacks).

> All of this are things that are especially useful when client side rendering (e.g. inside a wp-each), but also when setting values from the server side.

I'd appreciate it if you could test the new `wp-each` server directive processing out with your code to see if everything works as expected:

- https://github.com/WordPress/gutenberg/pull/58498

There's one thing that we don't support yet in the server, though.

We don't support passing a function like `isCurrentFilter` to generate the derived state dynamically based on the value of the context, like this JS code:

```js
store("myPlugin", {
  state: {
    get isCurrentFilter() {
      const { filter } = getContext();
      return filter.id === filter.name;
    },
  },
});
```

But the rest should work. You just have to do this:

```php
wp_interactivity_state("myPlugin", array(
  'filters' => array( ... ),
  'isCurrentFilter' => false // This needs to be static.
));
```

In the future, you will be able to do something like this:

```php
wp_interactivity_state("myPlugin", array(
  'filters' => array( ... ),
  'isCurrentFilter' => function() {
    // This can be dynamic.
    $context = wp_interactivity_get_context();
    return $context['filter']['id'] === $context['filter']['name'];
  }
));
```

Thank you! 🙏

_cc: @darerodz, it looks like supporting functions might be more important than we thought because of `wp-each`._

**Reply by gaambo** [reply]
Thank you very much for the detailed response and explanations. 🙏 
This all sounds very promising and well thought-through to me :) 
I guess I just have to wrap my head around some things that are different from AlpineJS and get used to it. Using the derived state getters and `getContext`/`getElement` in callbacks seems like more code first, but it totally makes sense that directly evaluating any directives will lead to problems (considering the amount of XSS security problems in the WordPress space). 
Using a negation operator is definitely a good point, that makes my code way more readable.

Maybe the table I'm using it fore at the moment also wasn't the best use case for that. I understand that server-side processing of some directives can make sense, and it can be a great way to add progressive-enhancement to such features. 

I'll play a little more with my implementation/demo and update accordingly here :)
